### PR TITLE
8225181: KeyStore should have a getAttributes method

### PR DESCRIPTION
**Pull request by wangweij**
Add `KeyStore::getAttributes` so that one can get the attributes of an entry without retrieving the entry first. This is especially useful for a private key entry which can only be retrieved with a password.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8225181](https://bugs.openjdk.java.net/browse/JDK-8225181): KeyStore should have a getAttributes method
 * [JDK-8275748](https://bugs.openjdk.java.net/browse/JDK-8275748): KeyStore should have a getAttributes method (**CSR**)

### Reviewers
 * [Sean Mullan](https://openjdk.java.net/census#mullan) (@seanjmullan - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6026/head:pull/6026` \
`$ git checkout pull/6026`

Update a local copy of the PR: \
`$ git checkout pull/6026` \
`$ git pull https://git.openjdk.java.net/jdk pull/6026/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6026`

View PR using the GUI difftool: \
`$ git pr show -t 6026`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6026.diff">https://git.openjdk.java.net/jdk/pull/6026.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/security/KeyStore.java [hunk 1]
```diff
@@ -1029,12 +1029,14 @@ public final String getType()
      * @param alias the alias name
      * @return an unmodifiable {@code Set} of attributes, possibly empty
      *      if the given alias does not exist, or there is no
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/KeyStore.java:1031]: s/is no/are no/

File: src/java.base/share/classes/java/security/KeyStore.java [hunk 2]
```diff
@@ -1029,12 +1029,14 @@ public final String getType()
      * @param alias the alias name
      * @return an unmodifiable {@code Set} of attributes, possibly empty
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/KeyStore.java:1030]: If the alias does not exist, then is it definitely always empty? I think the word "possibly" here may be a bit misleading. Maybe reword this as:
"an unmodifiable {@code Set} of attributes." This set is empty if the given alias does not exist or there are no attributes associated with the alias. This set may also be empty for PrivateKeyEntry or SecretKeyEntry entries that contain protected attributes and are only available through the ... "

File: src/java.base/share/classes/java/security/KeyStore.java (hunk 1 above)
Comment (wangweij) [reply src/java.base/share/classes/java/security/KeyStore.java]: Fixed.

File: src/java.base/share/classes/java/security/KeyStore.java (hunk 2 above)
Comment (wangweij) [reply src/java.base/share/classes/java/security/KeyStore.java]: Updated as suggested.

File: src/java.base/share/classes/java/security/KeyStore.java [hunk 3]
```diff
@@ -1020,6 +1020,35 @@ public final String getType()
         return this.type;
     }
 
+    /**
+     * Retrieves the attributes associated with the given alias.
+     *
+     * @implSpec
+     * The default implementation returns an empty {@code Set}.
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/KeyStore.java:1027]: Would it make more sense for the default impl to throw `UnsupportedOperationException` or maybe call `getEntry(alias, null)`? Otherwise, an application cannot know the difference between an alias that has no attributes and an alias that has attributes but is from a `KeyStore` impl that has not overridden the corresponding Spi method.

Comment (wangweij) [reply src/java.base/share/classes/java/security/KeyStore.java:1027]: The one benefit I can think of to throw a UOE is that the caller can fallback to `getEntry(...).getAttributes()` when an exception is thrown. However, as far as I know, our PKCS12KeyStore is the only KeyStore implementation that has made use of attributes. Therefore it's still not late for another implementation to start supporting both at the same time. For most of the KeyStore implementations, both `ks.getAttributes()` and `ks.getEntry(...).getAttributes()` returning empty seems more consistent.

Comment (seanjmullan) [reply src/java.base/share/classes/java/security/KeyStore.java:1027]: But we could just override those other implementations to always return an empty Set. I would prefer if an application could distinguish between these two cases w/o knowing whether the underlying implementation supports attributes or not.

**Comment by wangweij** [comment]
New commit pushed. Typo, I meant "refine".

**Comment by wangweij** [comment]
New commits pushed. Except for the merge commits, the only change is method spec update to match the text in the approved CSR.

**Review by seanjmullan — APPROVED** [review]
